src/gen.rs

```diff
@@ -430,6 +430,10 @@ fn gen_method_item(
     // Generate the list of argument used to call the method.
     let args = get_arg_list(sig.decl.inputs.iter())?;
 
+    // Determines the number of generic parameters, and what they are
+    let type_param_count = sig.decl.generics.type_params().count();
```

Comment:
I don't actually think we have to distinguish between 0 or more type parameters. Rust accepts empty parameter lists like `foo::<>()`. [Playground](https://play.rust-lang.org/?gist=ac088346fe004e626405f95f34f78307&version=stable&mode=debug&edition=2015). Unless I'm missing something, I think we can drop this distinction.

Reply:
I initially just tried adding it as such, but it seems that some field expressions will not accept generic types, meaning that some expansions will fail in that case.

Reply:
Ah, now I get it. So the problem is that an empty "type generics" from syn is an empty token stream and not `<>`. This means that we generate `self.method::(args)` when we don't have generics. That's incorrect syntax, hence the error.

But I found the method [`TypeGenerics::as_turbofish()`](https://docs.rs/syn/0.15.7/syn/struct.TypeGenerics.html#method.as_turbofish). This adds the double-colon if the token tree is not empty. So with that, we don't need that distinction. So what you basically need to do:

- Write `let generic_types = generic_types.as_turbofish();` in the line below the `split_for_impl` line.
- Remove the `::` from the `quote!`s, like this: `quote! { #proxy_ty_param::#name #generic_types(#args) }` (because if we don't have generics `#generic_types` will be an empty token stream, otherwise `::<T ...>`)
- Then you can remove this `count()` line and the `match`-distinctions below.